std/shared/src/main/scala-2/AsyncAwait.scala

```diff
@@ -0,0 +1,189 @@
+                  owner: Symbol,
+                  method: DefDef,
+                  awaitSymbol: Symbol,
+                  config: Map[String, AnyRef]
+              ): DefDef
+            }]
+            .markForAsyncTransform(
+              c.internal.enclosingOwner,
+              t,
+              awaitSym,
+              Map.empty
+            )
+        }
+        val name = TypeName("stateMachine$async")
+        // format: off
+        q"""
+          final class $name(dispatcher: _root_.cats.effect.std.Dispatcher[${c.prefix}._AsyncContext], callback: _root_.cats.effect.std.AsyncAwaitDsl.Callback[${c.prefix}._AsyncContext]) extends _root_.cats.effect.std.AsyncAwaitStateMachine(dispatcher, callback) {
+            ${mark(q"""override def apply(tr$$async: _root_.cats.effect.std.AsyncAwaitDsl.AwaitOutcome[${c.prefix}._AsyncContext]): _root_.scala.Unit = ${body}""")}
+          }
+          ${c.prefix}._AsyncInstance.flatten {
+            _root_.cats.effect.std.Dispatcher[${c.prefix}._AsyncContext].use { dispatcher =>
+              ${c.prefix}._AsyncInstance.async_[${c.prefix}._AsyncContext[AnyRef]](cb => new $name(dispatcher, cb).start())
+            }
+          }.asInstanceOf[${c.macroApplication.tpe}]
+        """
+      } catch {
+        case e: ReflectiveOperationException =>
+          c.abort(
+            c.macroApplication.pos,
+            "-Xasync is provided as a Scala compiler option, but the async macro is unable to call c.internal.markForAsyncTransform. " + e.getClass.getName + " " + e.getMessage
+          )
+      }
+  }
+
+}
+
+abstract class AsyncAwaitStateMachine[F[_]](
+    dispatcher: Dispatcher[F],
+    callback: AsyncAwaitDsl.Callback[F]
+)(implicit F: Async[F]) extends Function1[AsyncAwaitDsl.AwaitOutcome[F], Unit] {
+
+  // FSM translated method
+  //def apply(v1: AsyncAwaitDsl.AwaitOutcome[F]): Unit = ???
+
+  // Resorting to mutation to track algebraic product effects (like WriterT),
+  // since the information they carry would otherwise get lost on every dispatch.
+  private[this] var summary : F[Unit] = F.unit
+
+  private[this] var state$async: Int = 0
+
+  /** Retrieve the current value of the state variable */
+  protected def state: Int = state$async
+
+  /** Assign `i` to the state variable */
+  protected def state_=(s: Int): Unit = state$async = s
+
+  protected def completeFailure(t: Throwable): Unit =
+    callback(Left(t))
+
+  protected def completeSuccess(value: AnyRef): Unit = {
+    callback(Right(F.as(summary, value)))
+  }
+
+  protected def onComplete(f: F[AnyRef]): Unit = {
+    dispatcher.unsafeRunAndForget {
+      // Resorting to mutation to extract the "happy path" value from the monadic context,
+      // as inspecting the Succeeded outcome using dispatcher is risky on algebraic sums,
+      // such as OptionT, EitherT, ...
+      var awaitedValue: Option[AnyRef] = None
+      (summary *> f).flatTap(r => F.delay{awaitedValue = Some(r)}).start.flatMap(_.join).flatMap {
```

Comment:
It actually surprises me that scalafmt gives this line a pass.

```suggestion
      (summary *> f).flatTap(r => F.delay { awaitedValue = Some(r) }).start.flatMap(_.join).flatMap {
```

Also, why is the `.start.flatMap(_.join)` necessary? Why not just `guaranteeCase`? (is this just working around #2013?) At the very least we should use `uncancelable` to ensure the `Fiber` doesn't leak.

Reply:
> It actually surprises me that scalafmt gives this line a pass.

I had disabled formatting before the quasiquotes and forgotten to re-enable it. Fixed in 8971556

> Also, why is the .start.flatMap(_.join) necessary? 

2 birds, one stone : 
* working around #2013 as you guessed 
* prevent cases like OptionT.none to be run through the Dispatcher, which makes me nervous (even when calling unsafeRunAndForget). This is what I was pointing [here](https://github.com/typelevel/cats-effect/pull/1924#discussion_r641373346). When #2013 is fixed, could be replaced by `.forceR(F.void)` I suppose.

Reply:
> At the very least we should use uncancelable to ensure the Fiber doesn't leak.

Is this correct : 0220ebf ?

Reply:
Almost! Commented on the commit

Reply:
> This is what I was pointing here. When #2013 is fixed, could be replaced by .forceR(F.void) I suppose

I'm not sure I understand why the `forceR`? Asynchronous non-terminating fibers don't take any resources and are eventually GC'd when no longer referenced, so they aren't really a problem per se.

Reply:
> Asynchronous non-terminating fibers don't take any resources 

From the point of view of someone who doesn't know the innards of Dispatcher and how it relates to the effect's runtime, protecting against potential non-terminating behaviour seemed like a good reflex 😄. Also Dispatcher being polymorphic, can you guarantee that it'll be the case for all effect types ? (I reckon it'd be pretty bad if it wasn't the case, but still).

Reply:
> can you guarantee that it'll be the case for all effect types ? (I reckon it'd be pretty bad if it wasn't the case, but still).

I *think* that any `F` for which non-termination in `Dispatcher[F]` is problematic would also be an `F` for which *fibers* are problematic. I think.